### PR TITLE
Breakpoints

### PR DESCRIPTION
**Pull request by yann300**

### CODE REVIEW

File: src/app/debugger.js [hunk 1]
```diff
@@ -16,19 +16,51 @@ function Debugger (id, editor, compiler, executionContextEvent, switchToFile, of
...
+        break
+      }
+    }
+    source.row = row
+    return source
+  }
+
+  editor.event.register('breakpointCleared', (fileName, row) => {
+    if (self.compiler.lastCompilationResult.data) {
+      this.breakPointManager.remove(convertSourceLocation(this, fileName, row))
```
Comment (chriseth) [inline src/app/debugger.js]: The breakpoints should be file name based, the index is only valid for a single compilation run, but the breakpoints are not recreated.

File: src/app/debugger.js [hunk 2]
```diff
@@ -61,21 +93,44 @@ Debugger.prototype.debug = function (txHash) {
   })
 }
 
+Debugger.prototype.switchFile = function (rawLocation) {
+  var name = this.editor.getCacheFile() // current opened tab
+  var source = this.compiler.lastCompilationResult.data.sourceList[rawLocation.file] // auto switch to that tab
```
Comment (chriseth) [inline src/app/debugger.js]: what if lastCompilationResult does not exist?

File: src/app/editor.js [hunk 3]
```diff
@@ -17,6 +17,31 @@ function Editor (doNotLoadStorage, storage) {
...
 
+  var self = this
+  editor.on('guttermousedown', function (e) {
+    var target = e.domEvent.target
+    if (target.className.indexOf('ace_gutter-cell') === -1) {
+      return
+    }
+    var row = e.getDocumentPosition().row
+    var brkpoints = e.editor.session.getBreakpoints()
+    for (var k in brkpoints) {
```
Comment (chriseth) [inline src/app/editor.js]: please change to `breakpoints`

**Comment by axic** [comment]
Needs to be rebased.

**Pull request merged**